File: python/pygame/ball.py
import random
import pygame


class Ball:
    radius = 10
    pos = pygame.Vector2()
    vel = pygame.Vector2()

    def __init__(self, screen):
        self.reset(screen)
    
    def reset(self, screen):
        self.pos.x = screen.get_width() // 2
        self.pos.y = screen.get_height() // 2

        self.speed = 4
        
        if self.vel.x < 0:
            self.vel.x = self.speed
        else:
            self.vel.x = -self.speed

        self.bounce()
    
    def bounce(self):
        self.vel.y = random.choice([self.speed // 2,  -self.speed // 2])

    def draw(self, screen):
        pygame.draw.circle(screen, "white", self.pos, self.radius)

    def out_to_right(self, screen):
        return self.pos.x - self.radius > screen.get_width()
    
    def out_to_left(self):
        return self.pos.x + self.radius < 0
    
    def update(self, screen):
        self.pos.x += self.vel.x
        self.pos.y += self.vel.y

        if self.pos.y + self.radius > screen.get_height():
            self.vel.y = -self.speed // 2
        
        if self.pos.y - self.radius < 0:
            self.vel.y = self.speed // 2

File: java/swing/src/swing/Ball.java
package swing;

import java.awt.Color;
import java.awt.Graphics;
import java.util.Random;

public class Ball {
	public int x = 0;
	public int y = 0;
	public int vx = 0;
	public int vy = 0;
	public int speed = 4;
	public int radius = 10;

	public Ball(int screenWidth, int screenHeight) {
		reset(screenWidth, screenHeight);
	}

	public void reset(int screenWidth, int screenHeight) {
		x = screenWidth / 2;
		y = screenHeight / 2;
		speed = 4;

		if (vx < 0) {
			vx = speed;
		} else {
			vx = -speed;
		}

		bounce();
	}

	public void bounce() {
		Random rand = new Random();
		int choice = rand.nextInt(2);

		vy = choice == 0 ? -speed / 2 : speed / 2;
	}

	public void update(int screenWidth, int screenHeight) {
		x += vx;
		y += vy;
	}

	public void draw(Graphics g) {
		g.setColor(Color.WHITE);
		g.fillOval(x - radius, y - radius, radius * 2, radius * 2);
	}

	public Boolean outToLeft() {
		return x + radius < 0;
	}

	public Boolean outToRight(int screenWidth) {
		return x - radius > screenWidth;
	}
}
